presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/LogicalCteOptimizer.java

```diff
@@ -132,34 +139,68 @@ public boolean isPlanRewritten()
         }
     }
 
-    public class CteConsumerTransformer
+    public static class CteConsumerTransformer
             extends SimplePlanRewriter<CteTransformerContext>
     {
         private final PlanNodeIdAllocator idAllocator;
 
         private final VariableAllocator variableAllocator;
 
-        public CteConsumerTransformer(PlanNodeIdAllocator idAllocator, VariableAllocator variableAllocator)
+        private final Session session;
+
+        private static final List<Class<? extends PlanNode>> PRECOMPUTE_PLAN_NODES = ImmutableList.of(JoinNode.class, SemiJoinNode.class, AggregationNode.class);
+
+        private static final List<Class<? extends PlanNode>> DATA_SOURCES_PLAN_NODES = ImmutableList.of(TableScanNode.class, RemoteSourceNode.class, CteConsumerNode.class);
+
+        public CteConsumerTransformer(Session session, PlanNodeIdAllocator idAllocator, VariableAllocator variableAllocator)
         {
             this.idAllocator = requireNonNull(idAllocator, "idAllocator must not be null");
             this.variableAllocator = requireNonNull(variableAllocator, "variableAllocator must not be null");
+            this.session = requireNonNull(session, "Session is null");
         }
 
         @Override
         public PlanNode visitCteReference(CteReferenceNode node, RewriteContext<CteTransformerContext> context)
         {
-            context.get().addDependency(node.getCteName());
-            context.get().pushActiveCte(node.getCteName());
-            // So that dependent CTEs are processed properly
-            PlanNode actualSource = context.rewrite(node.getSource(), context.get());
-            context.get().popActiveCte();
-            CteProducerNode cteProducerSource = new CteProducerNode(node.getSourceLocation(),
-                    idAllocator.getNextId(),
-                    actualSource,
-                    node.getCteName(),
-                    variableAllocator.newVariable("rows", BIGINT), node.getOutputVariables());
-            context.get().addProducer(node.getCteName(), cteProducerSource);
-            return new CteConsumerNode(node.getSourceLocation(), idAllocator.getNextId(), actualSource.getOutputVariables(), node.getCteName());
+            HashMap<String, CTEInformation> cteInformationMap = session.getCteInformationCollector().getCteInformationMap();
+            CTEInformation cteInfo = cteInformationMap.get(node.getCteId());
+            switch (getCteMaterializationStrategy(session)) {
+                case HEURISTIC_COMPLEX_QUERIES_ONLY:
+                    if (!PlanNodeSearcher.searchFrom(node).where(planNode ->
```

Comment:
perhans refactor this expression into a function containsNodeType(node, NodeTypesList)

also: should we recursively handle CTEConsumers here? Right now we allow this if we see at least one CTE consumer as data source, but the cte producer for that one may be a trivial one (like a table scan)?

Reply:
Actually heuristic strategy flow for dependent CTEs is also not exactly accurate right now (eg: if a inner cte having a join/agg is materialized, the outer cte using this inner one might not be complex). This would need a completly different implementation which we would need to address in the cost based stategy. I was wondering if we could keep heuristic strategy as simple as possible

Reply:
PlanNodeSearcher.searchFrom(node) searches the node with CTEReferences, those are not rewritten as CteConsumers when this search happens

Reply:
If we visit the children of this node first and convert them to cte producers that might fix the problem. 
Also` Right now we allow this if we see at least one CTE consumer as data source, but the cte producer for that one may be a trivial one (like a values node)?` will never happen right? Since the heuristic strategy will ensures that the inner cte will never be converted to a cte producer.

Reply:
I fixed this test case in the refactor